docroot/modules/custom/sitenow_p2lb/src/Plugin/Action/RevertToParagraphs.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+namespace Drupal\sitenow_p2lb\Plugin\Action;
+
+use Drupal\Core\Action\ActionBase;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\node\NodeInterface;
+
+/**
+ * Action to revert page nodes back to Paragraphs.
+ *
+ * @Action(
+ *   id = "sitenow_p2lb_revert",
+ *   label = @Translation("Revert from V3 to V2"),
+ *   type = "node",
+ *   confirm = TRUE,
+ * )
+ */
+class RevertToParagraphs extends ActionBase {
+
+  use StringTranslationTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function access($object, AccountInterface $account = NULL, $return_as_object = FALSE) {
+    /** @var \Drupal\node\NodeInterface $object */
+    $result = $object->access('update', $account, TRUE)
+      ->andIf($object->field_page_content_block->access('edit', $account, TRUE));
+
+    return $return_as_object ? $result : $result->isAllowed();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute(NodeInterface $entity = NULL) {
+
+    // Get the node id.
+    $nid = $entity->id();
+
+    // Load the protected revision from the protected version id in that revision's values.
+    $protected_revision_id = sitenow_p2lb_get_protected_revision_id($nid);
+
+    // Guard against not finding the protected revision id.
+    if (!$protected_revision_id) {
+      return FALSE;
+    }
+
+    // Get the protected revision.
+    /** @var \Drupal\node\NodeInterface $protected_revision */
+    $protected_revision = sitenow_p2lb_get_protected_revision($protected_revision_id);
+
+    // Guard against not finding the protected revision.
+    if (!$protected_revision) {
+      return FALSE;
+    }
+
+    // Get a string representation of the original revision's timestamp.
+    $original_revision_timestamp = date('m/d/Y - h:i A', $protected_revision->getRevisionCreationTime());
+
+    /** @var \Drupal\Core\Entity\RevisionableStorageInterface $node_storage */
+    $node_storage = \Drupal::entityTypeManager()->getStorage('node');
+
+    // Create a new revision from node storage.
+    $new_revision = $node_storage->createRevision($protected_revision);
+
+    // Set moderation state.
+    $protected_revision->set('moderation_state', 'published');
```

Comment:
I think this might be leftover from previous iterations? `$protected_revision` doesn't end up getting saved, and the `$new_revision` created has the same moderation state as the protected revision, leading it to match whether that had previously been published or draft. At first I thought this should set the new revision to published, but that doesn't work if the page had previously been unpublished.

I'm wondering what the desired moderation state should really be, though. In the current setup, I tried:

1. Convert an unpublished page to v3
2. Publish the page
3. Revert to v2

In this case, the reverted v2 page has a moderation state matching the previous protected revision (draft). But the v3 version remains published. We could instead grab the moderation state of the latest revision to determine if it should be draft or published, but that feels odd for the editor experience.

Right now I'm thinking this is likely the best way to set the moderation state (where reverting to the last v2 version also reverts to that moderation state). If a page were published as v3 and the editor needed to revert to v2, the option to immediately publish that version is always available.

Reply:
```suggestion
```

Reply:
In the above example, this can get kind of funky if you convert/revert multiple times, and have a combo of 1) starting with an unpublished v2 page and 2) publish a v3 page. After you've converted and reverted, the following occurs:

1. Draft v2 page, published v3 page
2. Convert -- draft (newly converted) v3 page
3. Revert -- draft v2 page, published v3 page, page remains in the 'converted' view
4. Revert again -- draft v2 page, published v3 page, page no longer in the 'converted' view.

It still might only be an issue with trying to convert/revert over multiple times, in which case, some oddity is fine, and the option to just publish the reverted page remains available. But I do want to take a more structured approach to testing the different draft/published options that editors may actually use.